smac_planner/README.md

```diff
@@ -0,0 +1,159 @@
+      max_on_approach_iterations: 1000  # maximum number of iterations to attempt to reach goal once in tolerance, 2D only
+      max_planning_time_ms: 2000.0      # max time in ms for planner to plan, smooth, and upsample. Will scale maximum smoothing and upsampling times based on remaining time after planning.
+      smooth_path: false                # Whether to smooth searched path
+      motion_model_for_search: "DUBIN"  # 2D Moore, Von Neumann; SE2 Dubin, Redds-Shepp
+      angle_quantization_bins: 72       # For SE2 node: Number of angle bins for search, must be 1 for 2D node (no angle search)
+      minimum_turning_radius: 0.20      # For SE2 node & smoother: minimum turning radius in m of path / vehicle
+      reverse_penalty: 2.1              # For Reeds-Shepp model: penalty to apply if motion is reversing, must be => 1
+      change_penalty: 0.20              # For SE2 node: penalty to apply if motion is changing directions, must be >= 0
+      non_straight_penalty: 1.05        # For SE2 node: penalty to apply if motion is non-straight, must be => 1
+      cost_penalty: 1.3                 # For SE2 node: penalty to apply to higher cost zones
+
+      smoother:
+        smoother:
+          w_curve: 30.0                 # weight to minimize curvature of path
+          w_dist: 0.0                   # weight to bind path to original as optional replacement for cost weight
+          w_smooth: 30000.0             # weight to maximize smoothness of path
+          w_cost: 0.025                 # weight to steer robot away from collision and cost
+          cost_scaling_factor: 10.0     # this should match the inflation layer's parameter
+
+        # I do not recommend users mess with this unless they're doing production tuning
+        optimizer:
+          max_time: 0.10                # maximum compute time for smoother
+          max_iterations: 500           # max iterations of smoother
+          debug_optimizer: false        # print debug info
+          gradient_tol: 1.0e-10
+          fn_tol: 1.0e-20
+          param_tol: 1.0e-15
+          advanced:
+            min_line_search_step_size: 1.0e-20
+            max_num_line_search_step_size_iterations: 50
+            line_search_sufficient_function_decrease: 1.0e-20
+            max_num_line_search_direction_restarts: 10
+            max_line_search_step_expansion: 50
+```
+
+## Topics
+
+| Topic           | Type              |
+|-----------------|-------------------|
+| unsmoothed_path | nav_msgs/Path     |
+
+
+## Install
+
+```
+sudo apt-get install ros-<ros2-distro>-smac-planner
+```
+
+## Etc (Important Side Notes)
+
+### Potential Fields
+
+Many users and default navigation configuration files I find are really missing the point of the inflation layer. While it's true that you can simply inflate a small radius around the walls, the _true_ value of the inflation layer is creating a consistent potential field around the entire map. 
```

Comment:
After skimming the paper, this planner isn't using the inflation layer as a proxy for Voronoi Fields, is it? It seems a true Voronoi Field would assign a non-zero cost for all points on a map, excluding discretized points of free space along Voronoi cell walls. Where as the inflation layer has a finite influence/reach into larger free spaces.

Reply:
We use the inflation layer as a potential field and since we know how its computed, we can back out needed gradient information. The use of specifically a voronoi field is largely unnecessary and the paper way overblows the impact.

Reply:
But this section has nothing to do with voronoi, this just has to do with people not correctly utilizing potential fields with the inflation layer needed by heuristic driven planners

Reply:
Ok. That explains why the global plan seems to hug the contour of the inflation layer when traversing sides of open spaces, per the images you included in your above comments.

Hypothetically, for a empty warehouse scenario, with staggered rows of pylons, would we see the planner zigzag when traversing down a row? As opposed to straight down the alley? 

```
o  o  o  o
 o  o  o  o
o  o  o  o
```

Latching to the nearest inflation boundary, rather than a voronoi wall in the center seems like it could effect human intuition of the robots path.

Reply:
> Ok. That explains why the global plan seems to hug the contour of the inflation layer when traversing sides of open spaces, per the images you included in your above comments.

100%, and the NavFn A* and all others will do the same. Dijkstras gets around that by creating it own potential field, but any heuristic driven planner is  going to have some cost-to-travel that isn't dependent on a cost and something that takes into account cost. The result is that it will always steer and ride alongside cost areas if most space is 0. But if you make a nice potential field across the map, which the inflation layer will do, then you see nice and smooth performance using the spaces more effectively. This isn't unique to SMAC or NavFn, its just a feature of any heuristic driven search where we take into account cost. 

> Hypothetically, for a empty warehouse scenario, with staggered rows of pylons, would we see the planner zigzag when traversing down a row? As opposed to straight down the alley?

It depends. I don't think there's a catch-all answer to that. If the goal is straight across the row, it'll go straight. If its offset, it depends. The penalty functions can also be turned for this kind of behavior as well. You could crank up the non-straight heuristic so that it'll try to go straight when possible and that multiplier on cost would likely make it turn to hit the offset before or after the pylons since the multiplier would be lower in a lower-cost zone before/after since the inflation is lower.

I twouldn't latch to the boundary if you inflated the space properly, which is why I specifically call that out.